scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/KeyValueTag.scala

```diff
@@ -0,0 +1,65 @@
+package io.iohk.scalanet.discovery.ethereum
+
+import cats.implicits._
+import java.nio.charset.StandardCharsets.UTF_8
+import scodec.bits.ByteVector
+import scala.util.{Try, Success, Failure}
+
+/** Key value pairs that get added to the local ENR record as well as used
+  * as a critera for accepting remote ENRs.
+  */
+trait KeyValueTag {
+
+  /** Add a key-value pair to the outgoing ENR record.
+    * Return None if this tag is used only for filtering.
+    */
+  def toAttr: Option[(ByteVector, ByteVector)]
+
+  /** Apply a filter on incoming ENR records. */
+  def toFilter: KeyValueTag.EnrFilter
+}
+
+object KeyValueTag {
+
+  /** Return either a rejection message or unit, to accept the ENR. */
+  type EnrFilter = EthereumNodeRecord => Either[String, Unit]
+
+  def toFilter(tags: List[KeyValueTag]): EnrFilter = {
+    val filters = tags.map(_.toFilter)
+    enr => filters.traverse(_(enr)).void
+  }
+
+  class StringEquals(key: String, value: String) extends KeyValueTag {
+    private val keyBytes =
+      EthereumNodeRecord.Keys.key(key)
+
+    private val valueBytes =
+      ByteVector(value.getBytes(UTF_8))
+
+    override val toAttr =
+      Some(keyBytes -> valueBytes)
+
+    override val toFilter = enr =>
+      enr.content.attrs.get(keyBytes) match {
+        case Some(`valueBytes`) =>
```

Comment:
Hm, okay, I though tit was kind of cool not to have to name the variable if we don't need them, and it could scale to more occurrences too. Can change it back.

Reply:
Done.